### ISSUE TITLE
creating objects in Materialize Cloud is slow

### ISSUE THREAD

**Issue by JLDLaughlin** [MaterializeInc/materialize#13772]
### Feature request

@frankmcsherry reported high latencies trying to create objects in Materialize Cloud yesterday:

```
materialize=> create table flags (x int, y int, flag int, user text);
create view distinct_flags as select distinct * from flags;
create view correct_flags as
select x, y, flag, user from distinct_flags
where get_byte(sha256((x::text || y::text)::bytes), 0) % 6 = flag;
CREATE TABLE
Time: 1249.588 ms (00:01.250)
CREATE VIEW
Time: 953.576 ms
CREATE VIEW
Time: 826.520 ms
materialize=> 
```

**Comment by JLDLaughlin** [comment]
cc @philip-stoev

**Comment by ggnall** [comment]
@frankmcsherry Is this still happening/reproducible?  been sitting here awhile.

**Comment by mjibson** [comment]
Yes, this is roughly still happening. We are waiting for some persist leaks to be fixed so cockroach cpu load goes down before we re-measure.

**Comment by philip-stoev** [comment]
CREATE CLUSTER is indeed slow and getting slower as more clusters are added, see the "instance limits" test in Nightly CI.

**Comment by benesch** [comment]
We had another report of this yesterday, this time from a customer. Porting over the context from #15411:

> A customer is experiencing high latency when trying to create a view. We should [enable tracing](https://www.notion.so/materialize/Logs-Traces-Metrics-in-Honeycomb-525208809baa41cf9639ed17ff618474#c5bf8124c347490ea5821d7f4219e8c0) and investigate.
>
> Internal context can be found [here](https://materializeinc.slack.com/archives/C02LJ9EJP9S/p1665684184288829).

**Comment by benesch** [comment]
Note that the customer latencies were north of 30s, so definitely something going wrong.

**Comment by philip-stoev** [comment]
30-second delays after creating a source before it becomes queryable would be indicative of:
- #13958

**Comment by benesch** [comment]
That's `SELECT`ing from the source, though, right? This was literally the latency of `CREATE VIEW`.

**Comment by philip-stoev** [comment]
Yes, the SELECT. The first SELECT after the CREATE hangs for 30 seconds in cloudtest.

**Comment by benesch** [comment]
We used this issue today for our live debugging session and managed to capture a trace of a very slow 4s `CREATE VIEW` command: https://ui.honeycomb.io/materialize/datasets/environmentd_traces/result/HAtK8DTkXU5/trace/pzT3TLSaG5y?fields%5B%5D=c_name&fields%5B%5D=c_service.name&span=0baa8649f514f568

The trace points at a (very) slow CockroachDB transaction as a culprit.

@mjibson—any additional thoughts?

**Comment by benesch** [comment]
Not overly so. There’s a Grafana dashboard that reports aggregate Cockroach
resource usage!

**Comment by ggnall** [comment]
Pausing until Cloud adds jaeger tracing in the next few days, which will help us solve.

**Comment by jkosh44** [comment]
I think this is somewhat reproducible locally. I ran a test where I created 1000 identical views. The first one took ~200ms and the 1000th one took ~1 second. So the creation time does seem somewhat dependent on the number of existing objects.

Anecdotally I've tried to create 1000 tables locally and it seems like the first couple hundred were fast, but by the time I got to around 500 each table would take longer than 1 second. Though I haven't run the table experiment in awhile.

**Comment by jkosh44** [comment]
I added a bunch of print lines to see where we are spending our time and this is the result (Note: some of these labels are really not descriptive and was just used because I couldn't think of anything better):
```
Plan create view took: 1.31696ms
generate view ops took: 15.21007ms
Starting catalog_transact
Collecting things to drop took: 2.306µs
Dropping timelines took: 31.085µs
Validating resource limits took: 2.472267ms
Collecting temporary ids took: 30.966µs
Getting storage took: 12.461µs
Starting transaction took: 51.269619ms
Cloning catalog state took: 27.732025ms
ensure no stable uses took: 8.84µs
Op bulk stuff took: 2.823µs
Item serialization took: 1.336µs
Insert into transaction took: 488.874612ms
audit log took: 615.97µs
action took: 813.782µs
Callback took: 13.492µs
Adding all batches took: 354.491845ms
Stash append took: 12.308174ms
Sending builtin updates took: 29.668253ms
Drops took: 972ns
Everything else took: 4.059µs
```

`Insert into transaction took: 488.874612ms` Corresponds to this function:
https://github.com/MaterializeInc/materialize/blob/7482a5f78f0e58a37a7d63af40bdfd585bf2b695/src/adapter/src/catalog/storage.rs#L1206-L1227
The underlying implementation is O(n) in the number of items in the catalog, though it's a bit scary that it can take this long to add a single item.

`Adding all batches took: 354.491845ms` Corresponds to this part of the stash transaction:
https://github.com/MaterializeInc/materialize/blob/7482a5f78f0e58a37a7d63af40bdfd585bf2b695/src/adapter/src/catalog/storage.rs#L1511-L1639
I forget exactly what the end batch is supposed to look like, but the only difference between that batch and the current contents of the stash is a single new item.

There's most likely additional slow downs in cloud related to network calls that aren't reproducible locally, but these would also be good to fix.

**Comment by benesch** [comment]
#15842 is the latest development!
